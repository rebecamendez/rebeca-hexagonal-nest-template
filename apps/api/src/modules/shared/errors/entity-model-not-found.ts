export class EntityModelNotFoundError extends Error {
  public constructor(message: string) {
    super(message);
    Object.setPrototypeOf(this, EntityModelNotFoundError.prototype);
  }
}
