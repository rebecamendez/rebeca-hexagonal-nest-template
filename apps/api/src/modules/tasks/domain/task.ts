export class Task {
  public id!: number;
  public title!: string;
  public description!: string;
}
