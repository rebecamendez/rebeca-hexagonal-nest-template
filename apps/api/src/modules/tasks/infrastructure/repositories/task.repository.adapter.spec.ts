
import { StartedTestContainer } from 'testcontainers';
import { DataSource } from 'typeorm';

import { TaskEntity } from 'modules/shared/database/entities/task.entity';
import { TaskEntityBuilder } from 'modules/shared/database/tests/task.entity.builder';
import { EntityModelNotFoundError } from 'modules/shared/errors/entity-model-not-found';
import { containerSetup } from 'tests/test-containers.setup';
import { TaskRepository } from '../../application/ports/task.repository';
import { TaskRepositoryAdapter } from './task.repository.adapter';

describe('An task repository', () => {
  let dataSource: DataSource;
  let container: StartedTestContainer;
  let repository: TaskRepository;

  let taskEntity!: TaskEntity;

  const prepareScenario = async (dataSource: DataSource): Promise<void> => {
    taskEntity = await new TaskEntityBuilder().mock().save(dataSource);
  };

  beforeAll(async () => {
    ({ container, dataSource } = await containerSetup('task-repository-test'));

    repository = new TaskRepositoryAdapter(dataSource);

    await prepareScenario(dataSource);
  });

  afterAll(async () => {
    await dataSource.destroy();
    await container.stop();
  });

  describe('when getting tasks', () => {
    it('should return all tasks', async () => {
      const tasks = await repository.getTasks();
      expect(tasks).toMatchSnapshot();
    });
  });

  describe('when getting a task by id', () => {
    it('should return the task', async () => {
      const task = await repository.getTask(taskEntity.id);
      expect(task).toMatchSnapshot();
    });
  });

  describe('when creating a task', () => {
    it('should create a task', async () => {
      const task = await repository.createTask('title created', 'description created');
      expect(task).toMatchSnapshot();
    });
  });

  describe('when updating a task', () => {
    it('should update a task', async () => {
      const task = await repository.updateTask(taskEntity.id, 'title updated', 'description updated');
      expect(task).toMatchSnapshot();
    });
  });

  describe('when deleting a task', () => {
    it('should delete a task', async () => {
      await repository.deleteTask(taskEntity.id);
      await expect(repository.getTask(taskEntity.id)).rejects.toThrow(EntityModelNotFoundError);
    });
  });
});
