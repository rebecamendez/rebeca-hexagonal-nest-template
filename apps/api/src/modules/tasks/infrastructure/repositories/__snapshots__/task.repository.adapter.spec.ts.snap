// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`An task repository when creating a task should create a task 1`] = `
Task {
  "description": "description created",
  "id": 2,
  "title": "title created",
}
`;

exports[`An task repository when getting a task by id should return the task 1`] = `
Task {
  "description": "description",
  "id": 1,
  "title": "title",
}
`;

exports[`An task repository when getting tasks should return all tasks 1`] = `
[
  Task {
    "description": "description",
    "id": 1,
    "title": "title",
  },
]
`;

exports[`An task repository when updating a task should update a task 1`] = `
Task {
  "description": "description updated",
  "id": 1,
  "title": "title updated",
}
`;
